openflow-sdn/openflow-protocol.md

```diff
@@ -0,0 +1,65 @@
+The switch will generally match certain parameters like IP address, port number, MAC address, VLAN ID, etc. and select the best matching flow entry from the table and execute the action associated with that entry. Actions could be to drop the packet, forward it out a different port, flood the packet or send it to the controller for further inspection of the packet.
+
+If a switch does not have an entry for a packet, the switch might have a default entry or "TABLE_MISS" entry. This entry has the lowest priority and the actions can either be to drop the packet or send it to the controller. When the controller receives this kind of packet from a switch, it sends it to the application running at the application layer which can process the packet and let the controller know if a new flow entry needs to be inserted in the switch's flow table. If that's the case, the controller will insert a flow entry on the switch. 
+
+The next packet of the same kind will be dealt with by the switch at the Data layer as it already has an entry and appropriate actions would be taken. This improves the efficiency of the network by a huge factor. 
+
+## Advantages of Using OpenFlow
+There are several advantages of OpenFlow and SDN rather than traditional networks.
+1. SDN enables separation of control and data plane which means switches can use all their hardware resources in just forwarding data instead of computing routes. 
+2. OpenFlow provides an easy way of communication between controller and switch which can be easily implemented in an existing network. 
+3. Most current devices in the field support OpenFlow, it is not enabled by default, but we can easily enable and use it in the transition to SDN. 
+4. It provides security with a TLS connection to prevent snooping and DoS attacks on the controller and/or the network. 
+5. OpenFlow does **NOT** change the configuration for a switch. It just updates the flow tables which define the path for a packet.
+
+## Alternatives and other means of SDN 
+
+Although it seems that OpenFlow is ingrained with SDN, it is just an industry-standard practice to easily implement SDN in an organization. There are several other means with which one can program the SDN switches and insert flow entries in its flow tables.
+
+1. REST APIs: REST APIs are the most popular choice after OpenFlow to communicate with the switches. All the controller applications also have a REST module supported with them. The user can query the database of the switch to POST or GET data. "curl" is one of the most popular tools for this. An example of a REST call on the RYU controller to get all the switches in your network would be
+``` curl -XGET http://localhost:8080/stats/switches```
+
+2. Management protocols: Protocols like SSH, NETCONF, and SNMP could also be used to manage an OF enabled switch from the controller. This method is generally used for old devices that do not support OpenFlow. 
+
```

Comment:
```suggestion
Although it seems that OpenFlow is ingrained with SDN, it is just an industry-standard practice to implement SDN in an organization efficiently. There are several other means to program the SDN switches and insert flow entries in its flow tables.

1. **REST APIs:** REST APIs are the most popular choice after OpenFlow to communicate with the switches. All the controller applications also have a REST module supported with them. The user can query the switch database to POST or GET data. "Curl" is one of the most popular tools for this. An example of a REST call on the RYU controller to get all the switches in your network would be
``` curl -XGET http://localhost:8080/stats/switches```
2. **Management protocols:** Protocols like SSH, NETCONF, and SNMP could also manage an OF enabled switch from the controller. This method is generally used for old devices that do not support OpenFlow. 
```

Please also add an external link to "Curl".